go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,467 @@
+type DiskCache struct {
+	root             string         // path to the root directory of the disk cache.
+	maxCapacityBytes uint64         // if disk size exceeds this, old items will be evicted as needed.
+	mu               sync.Mutex     // protects the queue.
+	store            sync.Map       // map of keys to qitems.
+	queue            *priorityQueue // keys by last accessed time.
+	ctx              context.Context
+	shutdown         chan bool
+	shutdownOnce     sync.Once
+	gcTick           uint64
+	gcReq            chan uint64
+	gcDone           chan bool
+	stats            *DiskCacheStats
+}
+
+type DiskCacheStats struct {
+	TotalSizeBytes         int64
+	TotalNumFiles          int64
+	NumFilesStored         int64
+	TotalStoredBytes       int64
+	NumFilesGCed           int64
+	TotalGCedSizeBytes     int64
+	NumCacheHits           int64
+	NumCacheMisses         int64
+	TotalCacheHitSizeBytes int64
+	InitTime               time.Duration
+	TotalGCTime            time.Duration
+	TotalGCDiskOpsTime     time.Duration
+}
+
+func New(ctx context.Context, root string, maxCapacityBytes uint64) (*DiskCache, error) {
+	res := &DiskCache{
+		root:             root,
+		maxCapacityBytes: maxCapacityBytes,
+		ctx:              ctx,
+		queue: &priorityQueue{
+			items: make([]*qitem, 1000),
+		},
+		gcReq:    make(chan uint64, maxConcurrentRequests),
+		shutdown: make(chan bool),
+		gcDone:   make(chan bool),
+		stats:    &DiskCacheStats{},
+	}
+	start := time.Now()
+	defer func() { atomic.AddInt64((*int64)(&res.stats.InitTime), int64(time.Since(start))) }()
+	heap.Init(res.queue)
+	if err := os.MkdirAll(root, os.ModePerm); err != nil {
+		return nil, err
+	}
+	// We use Git's directory/file naming structure as inspiration:
+	// https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#:~:text=The%20subdirectory%20is%20named%20with%20the%20first%202%20characters%20of%20the%20SHA%2D1%2C%20and%20the%20filename%20is%20the%20remaining%2038%20characters.
+	eg, eCtx := errgroup.WithContext(ctx)
+	for i := 0; i < 256; i++ {
+		prefixDir := filepath.Join(root, fmt.Sprintf("%02x", i))
+		eg.Go(func() error {
+			if eCtx.Err() != nil {
+				return eCtx.Err()
+			}
+			if err := os.MkdirAll(prefixDir, os.ModePerm); err != nil {
+				return err
+			}
+			return filepath.WalkDir(prefixDir, func(path string, d fs.DirEntry, err error) error {
+				// We log and continue on all errors, because cache read errors are not critical.
+				if err != nil {
+					return fmt.Errorf("error reading cache directory: %w", err)
+				}
+				if d.IsDir() {
+					return nil
+				}
+				subdir := filepath.Base(filepath.Dir(path))
+				k, err := getKeyFromFileName(subdir + d.Name())
+				if err != nil {
+					return fmt.Errorf("error parsing cached file name %s: %w", path, err)
+				}
+				info, err := d.Info()
+				if err != nil {
+					return fmt.Errorf("error getting file info of %s: %w", path, err)
+				}
+				it := &qitem{
+					key: k,
+					lat: fileInfoToAccessTime(info),
```

Comment:
Why use the last access time instead of last modified time?
I don't believe that using the file system to track this information is a good idea. My implementations saved this as a dumb json file and that was very simple, I don't care about the actual time stamp, just the ordering; thus the file is a flat list of files in LRU for eviction order. This also removes the need for the priorityQueue type.